src/common/io_exerciser/Model.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include "IoOp.h"
+
+#include <boost/asio/io_context.hpp>
+
+#include "librados/librados_asio.h"
+
+#include "include/interval_set.h"
+#include "global/global_init.h"
+#include "global/global_context.h"
+#include "common/Thread.h"
+
+/* Overview
+ *
+ * class Model
+ *   Virtual class. Models apply IoOps generated by an
+ *   IoSequence, they can choose how many I/Os to execute in
+ *   parallel and scale up the size of I/Os by the blocksize
+ *
+ */
+
+namespace ceph {
+  namespace io_exerciser {
+
+    class Model
+    {
+    protected:
+      int num_io;
+      const std::string oid;
+      uint64_t block_size;
+
+    public:
+      Model(const std::string oid, uint64_t block_size);
```

Comment:
'const string' as a parameter spec does not make good sense.

Reply:
Basically - for all but string, accept as const T&.
If expecting a string - and planning to keep a copy - accept a string by value.
(and - if the environment is right - string_view)

Reply:
Was intending to put const T& but missed the &, got spot.

It sounds like by what you are saying however is that I should pass the string by value still, just without the const, is this correct? (I presume then move into the member variable instead of copy if this is what you are suggesting).

Reply:
Basically - we can usually count on the passed string to exist in the caller context until the called function returns. If that function does not stash a ref to the passed string somewhere (which will require it to outlive the caller),
than the best way to 'receive' it is as a string_view if possible (4 years ago - it would have been 'const string&').
If  it does require the string to outlive the caller: we'll need a copy (or move the string in, if the caller does not need it anymore). If the latter - that would usually require passing std::move(s).

Reply:
Ok, it sounds like because this is a constructor and it's stashing the oid in this case then I'm requiring the string to outlive the caller. I also want the oid to exist in the caller context after I construct this object potentially, so I do not want to std::move it out of the caller so it can still be used there - which leaves me with the option of passing it by copy.

In this case I would do this as passing it as "const std::string**&** oid", and then using the copy constructor of std::string in the initialiser list to copy this referenced string into the Model object so that the copy stored in Model can outlive the original string if necessary, using ", oid(oid)" in the initialiser list.

Have I understood this correctly?

Apologies for the questions, I'm just trying to nail down my understanding of the theory here for my knowledge.

Reply:
Yes. Sorry for nagging about that issue.